File: app/src/main/java/wxm/KeepAccount/utility/GlobalMsgHandler.java
package wxm.KeepAccount.utility;

import android.content.Intent;
import android.os.Handler;
import android.os.Message;
import android.util.Log;

import wxm.KeepAccount.define.EMsgType;
import wxm.androidutil.util.UtilFun;
import wxm.KeepAccount.define.GlobalDef;
import wxm.KeepAccount.define.UsrItem;


/**
 * global msg handler
 * Created by 123 on 2016/6/30.
 */
public class GlobalMsgHandler extends Handler {
    private static final String TAG = "GlobalMsgHandler";

    /**
     * reply msg
     * @param mh        receive handler
     * @param msg_type  origin msg type{@code arg1}）
     * @param msg_obj   object for reply{@code obj}
     */
    private static void ReplyMsg(Handler mh, int msg_type, Object msg_obj) {
        Message m = Message.obtain(mh, EMsgType.REPLAY.getId());
        m.arg1 = msg_type;
        m.obj = msg_obj;
        m.sendToTarget();
    }

    @Override
    public void handleMessage(Message msg) {
        Log.i(TAG, "receive msg : " + msg.toString());
        EMsgType et = EMsgType.getEMsgType(msg.what);
        if(null == et)
            return;

        switch (et) {
            case USR_ADD: {
                Object[] arr = UtilFun.cast(msg.obj);

                Intent data = UtilFun.cast(arr[0]);
                Handler h = UtilFun.cast(arr[1]);
                String usr = data.getStringExtra(UsrItem.FIELD_NAME);
                String pwd = data.getStringExtra(UsrItem.FIELD_PWD);

                if (ContextUtil.getUsrUtility().hasUsr(usr)) {
                    ReplyMsg(h, EMsgType.USR_ADD.getId(),
                            new Object[]{false, data, "用户已经存在！"});
                } else {
                    boolean ret = (null != ContextUtil.getUsrUtility().addUsr(usr, pwd));
                    ReplyMsg(h, EMsgType.USR_ADD.getId(),
                            new Object[]{ret, data});
                }
            }
            break;

            case USR_LOGIN: {
                Object[] arr = UtilFun.cast(msg.obj);

                Intent data = UtilFun.cast(arr[0]);
                Handler h = UtilFun.cast(arr[1]);

                String usr = data.getStringExtra(UsrItem.FIELD_NAME);
                String pwd = data.getStringExtra(UsrItem.FIELD_PWD);

                ContextUtil.setCurUsr(ContextUtil.getUsrUtility().CheckAndGetUsr(usr, pwd));
                boolean ret = (null != ContextUtil.getCurUsr());

                ReplyMsg(h, EMsgType.USR_LOGIN.getId(), ret);
            }
            break;

            case USR_LOGOUT: {
                ContextUtil.setCurUsr(null);
            }
            break;
        }
    }
}

File: app/src/main/java/wxm/KeepAccount/utility/GlobalMsgHandler.kt
package wxm.KeepAccount.utility

import android.content.Intent
import android.os.Handler
import android.os.Message
import org.greenrobot.eventbus.EventBus
import wxm.KeepAccount.db.LoginHistoryUtility

import wxm.KeepAccount.define.EMsgType
import wxm.androidutil.util.UtilFun
import wxm.KeepAccount.item.UsrItem
import wxm.KeepAccount.ui.login.DoLoginEvent
import wxm.androidutil.log.TagLog


/**
 * global msg handler
 * Created by WangXM on 2016/6/30.
 */
class GlobalMsgHandler : Handler() {
    override fun handleMessage(msg: Message) {
        TagLog.i("receive msg : " + msg.toString())
        val et = EMsgType.getEMsgType(msg.what) ?: return

        when (et) {
            EMsgType.USR_ADD -> {
                val arr = UtilFun.cast<Array<Any>>(msg.obj)

                val data = UtilFun.cast<Intent>(arr[0])
                val h = UtilFun.cast<Handler>(arr[1])
                val usr = data.getStringExtra(UsrItem.FIELD_NAME)
                val pwd = data.getStringExtra(UsrItem.FIELD_PWD)

                if (AppUtil.usrUtility.hasUsr(usr)) {
                    replyMsg(h, EMsgType.USR_ADD.id,
                            arrayOf(false, data, "用户已经存在！"))
                } else {
                    val ret = null != AppUtil.usrUtility.addUsr(usr, pwd)
                    replyMsg(h, EMsgType.USR_ADD.id,
                            arrayOf(ret, data))
                }
            }

            EMsgType.USR_LOGIN -> {
                val arr = UtilFun.cast<Array<Any>>(msg.obj)

                val data = UtilFun.cast<Intent>(arr[0])
                val h = UtilFun.cast<Handler>(arr[1])

                val usr = data.getStringExtra(UsrItem.FIELD_NAME)
                val pwd = data.getStringExtra(UsrItem.FIELD_PWD)

                AppUtil.curUsr = AppUtil.usrUtility.checkGetUsr(usr, pwd)
                val ret = null != AppUtil.curUsr

                replyMsg(h, EMsgType.USR_LOGIN.id, ret)
            }

            EMsgType.USR_LOGOUT -> {
                AppUtil.curUsr = null
                LoginHistoryUtility.cleanHistory()
                EventBus.getDefault().post(DoLoginEvent(this.javaClass.simpleName))
            }

            else -> {
                TagLog.e("can not handle msg : " + msg.toString())
            }
        }
    }

    companion object {
        /**
         * reply msg
         * @param mh        receive handler
         * @param msg_type  origin msg type`arg1`）
         * @param msg_obj   object for reply`obj`
         */
        private fun replyMsg(mh: Handler?, msg_type: Int, msg_obj: Any) {
            val m = Message.obtain(mh, EMsgType.REPLAY.id)
            m.arg1 = msg_type
            m.obj = msg_obj
            m.sendToTarget()
        }
    }
}

